pkg/operator/sync.go

```diff
@@ -301,6 +303,25 @@ func (optr *Operator) syncRenderConfig(_ *renderConfig) error {
 		}
 	}
 	spec.AdditionalTrustBundle = trustBundle
+	clusterversionCfg, err := optr.configClient.ConfigV1().ClusterVersions().Get(context.TODO(), "version", metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		glog.Infof("ClusterVersionConfig 'version' does not exist or has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	if proxy != nil && clusterversionCfg != nil {
+		icspRules, err := optr.icspLister.List(labels.Everything())
+		if err != nil && errors.IsNotFound(err) {
+			icspRules = []*apioperatorsv1alpha1.ImageContentSourcePolicy{}
+		} else if err != nil {
+			return err
+		}
+		if err := ctrlcommon.ProxyValidation(&proxy.Status, clusterversionCfg.Status.Desired.Version, icspRules); err != nil {
```

Comment:
I feel like you could talk me into putting this in a separate place for "sanity checks" before config rollout/eviction/drain, but I have concerns about the proposed location in `SyncRenderedconfig()` -- if the proxyconfig fails this test for whatever reason, we don't get a `RenderConfig`, which prevents the rest of the sync functions from running, and that is problematic for general cluster stability during normal operations (among other things, it affects certificate rotations). 

Specifically, in a case where the proxy was "valid" when it was configured, but is down/unreachable/etc at the time of the check, the MCO would degrade, wouldn't it? 

I don't know that I have a spot picked out where it *should* go, because the MCO has typically not accepted preflight checks of this nature, but I'm kind of sympathetic here   :smile:

Reply:
@yuqi-zhang could you help locate where MCO deploys the proxy settings to the Nodes so that the proxyvalidatoin can be done there?

Reply:
Right, so the exact details are a bit up to debate. The way it's set up in this PR right now is blocking at the `operator` level which is potentially dangerous for reasons John has listed, and I agree that we should probably think about moving this to a consolidated "checking" location.

So, if we want to achieve the point of "don't roll out the proxy to nodes unless it *most likely* works", then it likely will have to happen at the controller level, between:
1. https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/render/render_controller.go#L546, where the rendered MC is generated for a pool
2. https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/node/node_controller.go#L846, where the config is rolled out to the MCP

So then this would be something like `validateIncomingRenderedConfig` before it gets rolled out to the pool, somehow, where right now we just validate the proxy, but could be extended as a node-specific pre-flight check of some sort. We could even have a flag that enables/disables this, if we don't want to change default behaviour.

The other side of this issue is, as we move towards layering, what if I built a new format OS image with a proxy built into it somehow? This validation path would not catch that if done directly in the image.

One last extension thought on validation that's a bit more encompassing: have a (flag enabled?) option to create an extra canary node on incoming updates to see if that node would break, before upgrading the rest of the nodes. That's a bit too far though.

In summary, I think this is a pretty complex topic. Right now I think maybe the safest option is at the controller level, but how exactly that would be done is a bit up in the air

Reply:
> The other side of this issue is, as we move towards layering, what if I built a new format OS image with a proxy built into it somehow? This validation path would not catch that if done directly in the image.

Ultimately I think what we want is https://github.com/ostreedev/ostree/issues/2725 - basically, we try booting the new configuration, and roll back if kubelet isn't able to start.

Reply:
@yuqi-zhang Please review, need help with the operator code auto-generation regarding to the `func (f *fixture) newController()` (pkg/controller/render/render_controller_test.go)

Reply:
> need help with the operator code auto-generation regarding to the func (f *fixture) newController() (pkg/controller/render/render_controller_test.go)

Sorry, I don't quite follow, what is the issue with the code autogen?